### ISSUE TITLE
Optional property type is not detected in a generic function

### ISSUE THREAD

**Issue by anvish** [Microsoft/TypeScript#12436]
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.201xxxxx)
2.1.1
**Code**

```ts
interface State {
    mode?: 'a' | 'b'
}

function assign<T>(first: T, second: T) { }

const state: State = { mode: 'a' }

assign(state, { mode: 'b' }) // error

// [ts] The type argument for type parameter 'T' cannot be inferred from the usage. Consider specifying the type arguments explicitly.
//        Type argument candidate 'State' is not a valid type argument because it is not a supertype of candidate '{ mode: string; }'.
//          Types of property 'mode' are incompatible.
//            Type 'string' is not assignable to type '"a" | "b"'.
```
But if `?` is removed, type is detected correctly:
```ts
interface State {
    mode: 'a' | 'b'
}

function assign<T>(first: T, second: T) { }

const state: State = { mode: 'a' }

assign(state, { mode: 'b' }) // no error
```
**Expected behavior:**
Type is detected for both optional and not optional properties.
**Actual behavior:**
Type is detected for not optional properties only.

**Comment by JoshuaKGoldberg** [comment]
Bump, seeing what looks like a variant of this in 2.2.

https://github.com/DefinitelyTyped/DefinitelyTyped/issues/15768


```typescript
interface TestInterface {
    required: 'aggregate';
    // The error goes away if this is just a boolean
    other: string | boolean;
}

const actual: TestInterface = {
    required: 'aggregate',
    other: true
};

/*
The type argument for type parameter 'T' cannot be inferred from the usage. Consider specifying the type arguments explicitly.
  Type argument candidate 'TestInterface' is not a valid type argument because it is not a supertype of candidate '{ required: string; other: boolean; }'.
    Types of property 'required' are incompatible.
      Type 'string' is not assignable to type '"aggregate"'.
*/
assert.deepEqual(actual, {
    required: 'aggregate',
    other: true
});

// This is fine!
assert.deepEqual(actual, {
    required: 'aggregate'
});
```

**Comment by mhegazy** [comment]
Should be working in latest.
